File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/New/Heisenberg/Subsystems/Linkage.java
package org.firstinspires.ftc.teamcode.New.Heisenberg.Subsystems;

import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.HardwareMap;

import org.firstinspires.ftc.teamcode.New.Utilities.Controller;
import org.firstinspires.ftc.teamcode.New.Utilities.PIDParams;


public class Linkage {
    DcMotor linkage;

    public LinkageState linkageState;
    double goalPos;
    final double ticksPerRev = 28;

    Controller PIDcontroller = new Controller(new PIDParams(0,0,0,0));

    public Linkage (HardwareMap hardwareMap){
        linkage = hardwareMap.get(DcMotor.class, "clawFlipper");
    }

    public void init() {
        linkage.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
    }

    public void update(double goalDegrees) {

        double goalTicks = (goalDegrees/360) * ticksPerRev;

        switch (linkageState){
            //7 ticks = 0.25 revolutions = 90 degrees, so straight up
            case VERTICAL: goalPos = 7; break;
            case INPUT: goalPos = goalTicks; break;
        }


        int linkageEncoder = linkage.getCurrentPosition();
        double power = PIDcontroller.calculate(goalPos ,linkageEncoder);

        if(linkageState == LinkageState.IDLE){
            power = 0;
        }

        linkage.setPower(power);

    }

    public enum LinkageState {
        VERTICAL,
        INPUT,
        IDLE
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/New/Heisenberg/Subsystems/Linkage.kt
package org.firstinspires.ftc.teamcode.New.Heisenberg.Subsystems

import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.teamcode.New.Heisenberg.OpModes.Saturday.SlideRotator
import org.firstinspires.ftc.teamcode.New.SubSystems
import org.firstinspires.ftc.teamcode.New.Utilities.Controller
import org.firstinspires.ftc.teamcode.New.Utilities.PIDParams

class Linkage(hardwareMap: HardwareMap): SubSystems {
    val motor: DcMotorEx = hardwareMap.get(DcMotorEx::class.java,"linkage")

    init {
        motor.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        motor.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER
    }

    //1.27 mult

    private val controller = Controller(PIDParams(1.7, 0.0, .002, 0.4))
    enum class State(val angle: Double){
        Horizontal(15.0 + 90),
        Basket(80.0 + 90),
        Vertical(90.0+90),
        Specimin(45.0 +90),
        SubmersibleStart(11.0 + 90);
    }

    val mult = 1.27


    override var state = State.Horizontal

    override fun update() {
        val currentAngle  = (Math.PI /180) * ((((motor.currentPosition*2*11/12)/1425.05923061 * 2 * Math.PI) *180/ Math.PI)/4 + 90)
        motor.power  =  controller.calculate((state.angle* Math.PI/180*mult -currentAngle),currentAngle)
    }


}
